pyro/ops/stats.py

```diff
@@ -0,0 +1,215 @@
+        if m == 1:
+            return N
+        N += 1
+
+
+def autocorrelation(input, dim=0):
+    """
+    Computes the autocorrelation of samples at dimension `dim`.
+
+    Reference: https://en.wikipedia.org/wiki/Autocorrelation#Efficient_computation
+
+    :param torch.Tensor input: the input tensor.
+    :param int dim: the dimension to calculate autocorrelation.
+    :returns torch.Tensor: autocorrelation of `input`.
+    """
+    if (not input.is_cuda) and (not torch.backends.mkl.is_available()):
+        raise NotImplementedError("For CPU tensor, this method is only supported "
+                                  "with MKL installed.")
+
+    # Adapted from Stan implementation
+    # https://github.com/stan-dev/math/blob/develop/stan/math/prim/mat/fun/autocorrelation.hpp
+    N = input.size(dim)
+    M = _fft_next_good_size(N)
+    M2 = 2 * M
+
+    # transpose dim with -1 for Fourier transform
+    input = input.transpose(dim, -1)
+
+    # centering and padding x
+    centered_signal = input - input.mean()
+    pad = input.new_zeros(input.shape[:-1] + (M2 - N,))
+    centered_signal = torch.cat([centered_signal, pad], dim=-1)
+
+    # Fourier transform
+    freqvec = torch.rfft(centered_signal, signal_ndim=1, onesided=False)
+    # take square of magnitude of freqvec (or freqvec x freqvec*)
+    freqvec_gram = freqvec.pow(2).sum(-1, keepdim=True)
+    freqvec_gram = torch.cat([freqvec_gram, input.new_zeros(freqvec_gram.shape)], dim=-1)
+    # inverse Fourier transform
+    autocorr = torch.irfft(freqvec_gram, signal_ndim=1, onesided=False)
+
+    # truncate and normalize the result, then transpose back to original shape
+    autocorr = autocorr[..., :N]
+    autocorr = autocorr / input.new_tensor(range(N, 0, -1))
+    autocorr = autocorr / autocorr[..., :1]
+    return autocorr.transpose(dim, -1)
+
+
+def autocovariance(input, dim=0):
+    """
+    Computes the autocovariance of samples at dimension `dim`.
+
+    :param torch.Tensor input: the input tensor.
+    :param int dim: the dimension to calculate autocorrelation.
+    :returns torch.Tensor: autocorrelation of `input`.
+    """
+    return autocorrelation(input) * input.var(dim, unbiased=False, keepdim=True)
+
+
+def _cummin(input):
+    """
+    Computes cummulative minimum of input at dimension `dim=0`.
+
+    :param torch.Tensor input: the input tensor.
+    :returns torch.Tensor: accumulate min of `input` at dimension `dim=0`.
+    """
+    # FIXME: is there a better trick to find accumulate min of a sequence?
+    N = input.size(0)
+    input_tril = input.unsqueeze(0).repeat((N,) + (1,) * input.dim())
+    triu_mask = input.new_ones(N, N).triu(diagonal=1).reshape((N, N) + (1,) * (input.dim() - 1))
+    triu_mask = triu_mask.expand((N, N) + input.shape[1:]) > 0.5
+    input_tril.masked_fill_(triu_mask, input.max())
+    return input_tril.min(dim=1)[0]
+
+
+def effective_sample_size(input, sample_dim=0, chain_dim=1):
+    """
+    Computes effective sample size of input.
```

Comment:
I think we should also document what kind of input sizes / constraints we expect. e.g. if the latents have different shapes, this would not work. Are these diagnostics well behaved with multidimensional (but equally sized) latents, or is there an expectation that they will all be unidimensional?

Reply:
I try to make these methods have the same interface as `torch.foo(input, ...)`. This just accepts 1 input tensor. If you want to apply it for multiple latents, you have to put it in a loop. The other way is to flatten then merge all latents into a 1D tensor, then apply this function.

Reply:
I see; so input can be arbitrary shaped but the expectation is that it represents samples only from a single latent site.